### ISSUE TITLE
Disabling SNI doesn't work with java 8u144

### ISSUE THREAD

**Issue by nodefactory-bk** [akka/akka-http#1331]
Hi!

Just discovered that the options for disabling SNI don't seem to work when using java 8u144.
It worked just fine in j8u131

Based on the oracle changelog for 8u141: http://www.oracle.com/technetwork/java/javase/8u141-relnotes-3720385.html
I believe this has something to do with https://bugs.openjdk.java.net/browse/JDK-8144566

This is the code I use for creating the conenction context:
````
  private lazy val _httpsContext: HttpsConnectionContext =
  {
    val ctx = sslContext
    val params = ctx.getDefaultSSLParameters

    val cfg = akkaSSLConfig.mapSettings { s =>
      val loose = s.loose
        .withDisableSNI(true)
        .withAcceptAnyCertificate(true)
        .withDisableHostnameVerification(true)

      s.withLoose(loose)
    }

    new HttpsConnectionContext(ctx, Some(cfg), sslParameters = Some(params))
  }

````

I have tested it and it works great with openjdk 8u131 and oracle java 8u131 but fails with unrecognized_name (because SNI is enabled) on openjdk 8u144 and oracle java 8u144

**Comment by nodefactory-bk** [comment]
Based om further investiagion it seems it is no longer sufficient to just exclude hostnames from the SNI/hostname verification as is done in TLS.scala
It seems you need to create the ssl engine without hostname.
This patch seems to do it:
```

diff --git a/akka-stream/src/main/scala/akka/stream/scaladsl/TLS.scala b/akka-stream/src/main/scala/akka/stream/scaladsl/TLS.scala
index 7955bedd3a..c01977ee04 100644
--- a/akka-stream/src/main/scala/akka/stream/scaladsl/TLS.scala
+++ b/akka-stream/src/main/scala/akka/stream/scaladsl/TLS.scala
@@ -72,15 +72,18 @@ object TLS {
     sslConfig:    Option[AkkaSSLConfig],
     firstSession: NegotiateNewSession, role: TLSRole,
     closing: TLSClosing = IgnoreComplete, hostInfo: Option[(String, Int)] = None): scaladsl.BidiFlow[SslTlsOutbound, ByteString, ByteString, SslTlsInbound, NotUsed] = {
+
     def theSslConfig(system: ActorSystem): AkkaSSLConfig =
       sslConfig.getOrElse(AkkaSSLConfig(system))
 
     val createSSLEngine = { system: ActorSystem ⇒
+      val config = theSslConfig(system)
+
       val engine = hostInfo match {
-        case Some((hostname, port)) ⇒ sslContext.createSSLEngine(hostname, port)
-        case None                   ⇒ sslContext.createSSLEngine()
+        case Some((hostname, port)) if !config.config.loose.disableSNI ⇒ sslContext.createSSLEngine(hostname, port)
+        case _ ⇒ sslContext.createSSLEngine()
       }
-      val config = theSslConfig(system)
+
       config.sslEngineConfigurator.configure(engine, sslContext)
       engine.setUseClientMode(role == Client)
 
```

Doesn't seem to have any other problems on my tests.

Ps. Maybe move this ticket to akka since it isn't really akka-http ?

**Comment by jrudolph** [comment]
@nodefactory-bk from a quick glance it seems your observation is correct. Server names can be configured using `SSLParameters` but even if none is given there the one name from `createSSLEngine(host, port)` will be used. However, the change you propose would also disable hostname verification. While this seems to be intended for your usecase, it is insecure in general. I'm reluctant to make changes that weaken security (intentionally or unintentionally).

As a workaround you can globally disable SNI by adjusting the system property `jsse.enableSNIExtension` as [explained in the documentation](http://docs.oracle.com/javase/7/docs/technotes/guides/security/jsse/JSSERefGuide.html).

I understand that it is somewhat bad that ssl-config / akka-http suggests that disabling SNI would work and now doesn't any more but at this point I don't see us fixing it with the system property available (otherwise, you should always be able to fix the server to ignore SNI).

**Comment by nodefactory-bk** [comment]
Hi!
Are you sure this would disable hostname verification?
Java SSL uses separate hostname verifiers even allowing one to set custom ones, so forcing verification based on the host name provided in the ssl engine seems somewhat counter intuitive.
Especially since SNI and hostname verification really have nothing to do with one another in TLS.

I tested the hostname verification with my patch and SNI disabled and it seems to work just fine.
If I use the code below:
```
  private lazy val _httpsContext: HttpsConnectionContext =
  {
    val ctx = sslContext
    val params = ctx.getDefaultSSLParameters

    val cfg = akkaSSLConfig.mapSettings { s =>
      val loose = s.loose
        .withDisableSNI(true)

      s.withLoose(loose)
    }

    new HttpsConnectionContext(ctx, Some(cfg), sslParameters = Some(params))
  }
```
Which is identical to the snippet in my original post but with hostchecks and certificate checks enabled I get a connection with no SNI unrecognized_name failures and it properly rejects the host as not matching the one I requested. 
This is with akka patched with my patch. So unless I'm missing another case it seems my patch won't affect hostname verification.

Unfortunately it's not always possible to force someone else to configure their servers correctly, especially since this is actually a bug in java that is caused by java treating the "unrecognized_name" WARNING as a fatal condition causing connection problems even when the certificate is correct.
No browsers do this. And setting it globally doesn't work since I need to connect to servers with and without SNI from the same application.

If nothing else, changing my patch to require both disableSNI and disableHostnameVerification would be fine by me at least.

Or if there was some other way to influence the creation of the SSL engine, so I could provide my own parameters that'd be fine too.

**Comment by nodefactory-bk** [comment]
Hi!
I've been running my patch in production for about a month now.
I haven't noticed any ill effects. Certificate validation works as it should and I can use connections
with and without SNI from the same app.

**Comment by raboof** [comment]
@nodefactory-bk though this is something that needs to be carefully double-checked, your proposal generally seems reasonable to me (and indeed also requiring `disableHostnameVerification` would be pretty safe). Would you be interested in turning your patch into a pull request?

**Comment by nodefactory-bk** [comment]
Sure, here you go:
https://github.com/akka/akka/pull/23878

Not sure I did it correctly since I rarely use github for pull requests =)

**Comment by jrudolph** [comment]
Seems to have been fixed by akka/akka#23878?
